compute/src/main/java/org/jclouds/compute/domain/internal/ArbitraryCpuRamTemplateBuilderImpl.java

```diff
@@ -0,0 +1,84 @@
+ * limitations under the License.
+ */
+package org.jclouds.compute.domain.internal;
+
+import com.google.common.base.Supplier;
+import org.jclouds.collect.Memoized;
+import org.jclouds.compute.domain.Hardware;
+import org.jclouds.compute.domain.HardwareBuilder;
+import org.jclouds.compute.domain.Image;
+import org.jclouds.compute.domain.Processor;
+import org.jclouds.compute.domain.TemplateBuilder;
+import org.jclouds.compute.options.TemplateOptions;
+import org.jclouds.compute.util.AutomaticHardwareIdSpec;
+import org.jclouds.domain.Location;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static org.jclouds.compute.util.AutomaticHardwareIdSpec.AutomaticHardwareIdSpecBuilder;
+import static org.jclouds.compute.util.AutomaticHardwareIdSpec.isAutomaticId;
+import static org.jclouds.compute.util.AutomaticHardwareIdSpec.parseId;
+
+public class ArbitraryCpuRamTemplateBuilderImpl extends TemplateBuilderImpl {
+   @Inject
+   protected ArbitraryCpuRamTemplateBuilderImpl(@Memoized Supplier<Set<? extends Location>> locations,
+         @Memoized Supplier<Set<? extends Image>> images, @Memoized Supplier<Set<? extends Hardware>> hardwares,
+         Supplier<Location> defaultLocation, @Named("DEFAULT") Provider<TemplateOptions> optionsProvider,
+         @Named("DEFAULT") Provider<TemplateBuilder> defaultTemplateProvider) {
+      super(locations, images, hardwares, defaultLocation, optionsProvider, defaultTemplateProvider);
+   }
+
+   protected Hardware automaticHardwareForCpuAndRam(double cores, int ram) {
+      return new HardwareBuilder()
+            .id(AutomaticHardwareIdSpecBuilder(cores, ram).toString())
+            .ram(ram)
+            .processor(new Processor(cores, 1.0))
+            .build();
+   }
+
+   protected Hardware findHardwareWithId(Set<? extends Hardware> hardwaresToSearch) {
+      try {
+            return super.findHardwareWithId(hardwaresToSearch);
+      } catch (NoSuchElementException ex) {
+            if (isAutomaticId(hardwareId)) {
+               AutomaticHardwareIdSpec spec = parseId(hardwareId);
+               return automaticHardwareForCpuAndRam(spec.getCores(), spec.getRam());
+            }
+            else{
+               throw ex;
+            }
+        }
+    }
+
+   protected Hardware resolveHardware(Set<? extends Hardware> hardwarel, final Iterable<? extends Image> images) {
+      try {
+         return super.resolveHardware(hardwarel, images);
+      }
+      catch (NoSuchElementException ex) {
+         if (super.minCores != 0 && super.minRam != 0) {
+            return automaticHardwareForCpuAndRam(minCores, minRam);
+         }
+         else throw new NoSuchElementException("No hardware profile matching the given criteria was found. " +
+               "If you want to use exact values, please set the minCores and minRam values");
```

Comment:
Wrap the original exception as the cause of this one, to keep the complete stacktrace.